presto-pinot/src/main/java/com/facebook/presto/pinot/PinotPageSource.java

```diff
@@ -0,0 +1,392 @@
+        long startTimeNanos = System.nanoTime();
+        if (pinotColumnNameIndexMap.isEmpty()) {
+            int idx = 0;
+            for (PinotColumnHandle columnHandle : columnHandles) {
+                pinotColumnNameIndexMap.put(columnHandle.getColumnName(), idx++);
+            }
+        }
+        String pinotQuery = getPinotQuery(pinotConfig, columnHandles, split.getPinotFilter(), split.getTimeFilter(), split.getTableName(), split.getLimit());
+        Map<ServerInstance, DataTable> dataTableMap = pinotQueryClient.queryPinotServerForDataTable(pinotQuery, split.getHost(), split.getSegment());
+        dataTableMap.values().stream()
+                // ignore empty tables and tables with 0 rows
+                .filter(table -> table != null && table.getNumberOfRows() > 0)
+                .forEach(dataTable ->
+                {
+                    // Store each dataTable which will later be constructed into Pages.
+                    // Also update estimatedMemoryUsage, mostly represented by the size of all dataTables, using numberOfRows and fieldTypes combined as an estimate
+                    int estimatedTableSizeInBytes = IntStream.rangeClosed(0, dataTable.getDataSchema().size() - 1)
+                            .map(i -> getEstimatedColumnSizeInBytes(dataTable.getDataSchema().getColumnDataType(i)) * dataTable.getNumberOfRows())
+                            .reduce(0, Integer::sum);
+                    dataTableList.add(new PinotDataTableWithSize(dataTable, estimatedTableSizeInBytes));
+                    estimatedMemoryUsageInBytes += estimatedTableSizeInBytes;
+                });
+        ImmutableList.Builder<Type> types = ImmutableList.builder();
+        columnHandles.stream()
+                .map(PinotPageSource::getTypeForBlock)
+                .forEach(types::add);
+        columnTypes = types.build();
+        readTimeNanos += System.nanoTime() - startTimeNanos;
+        isPinotDataFetched = true;
+    }
+
+    @Override
+    public void close()
+    {
+        if (closed) {
+            return;
+        }
+        closed = true;
+    }
+
+    /**
+     * Generates the {@link com.facebook.presto.spi.block.Block} for the specific column from the {@link #currentDataTable}.
+     *
+     * <p>Based on the original Pinot column types, write as Presto-supported values to {@link com.facebook.presto.spi.block.BlockBuilder}, e.g.
+     * FLOAT -> Double, INT -> Long, String -> Slice.
+     *
+     * @param blockBuilder blockBuilder for the current column
+     * @param columnType type of the column
+     * @param columnIdx column index
+     */
+
+    private void writeBlock(BlockBuilder blockBuilder, Type columnType, int columnIdx)
+    {
+        Class<?> javaType = columnType.getJavaType();
+        DataSchema.ColumnDataType pinotColumnType = currentDataTable.getDataTable().getDataSchema().getColumnDataType(columnIdx);
```

Comment:
Having this implicit dependency is very not obvious. Instead of passing a columnIdx and fetching the data from the class level field - please pass the column value directly here (and if index is needed - you can do that as well)